ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/RegisterSchemaCallback.java

```diff
@@ -43,17 +44,66 @@ public void onDeserializationFailure(
   ) {
     final String sourceSubject = source + KsqlConstants.SCHEMA_REGISTRY_VALUE_SUFFIX;
     final String changelogSubject = changelog + KsqlConstants.SCHEMA_REGISTRY_VALUE_SUFFIX;
+
+    // all schema registry events start with a magic byte 0x0 and then four bytes
+    // indicating the schema id - we extract that schema id from the data that failed
+    // to deserialize and then register it into the changelog subject
+    final int id = ByteBuffer.wrap(data, 1, Integer.BYTES).getInt();
+    final SchemaRegisterEvent event = new SchemaRegisterEvent(id, sourceSubject, changelogSubject);
+
     try {
-      // all schema registry events start with a magic byte 0x0 and then four bytes
-      // indicating the schema id - we extract that schema id from the data that failed
-      // to deserialize and then register it into the changelog subject
-      final int id = ByteBuffer.wrap(data, 1, Integer.BYTES).getInt();
-
-      LOG.info("Trying to fetch & register schema id {} under subject {}", id, changelogSubject);
-      final ParsedSchema schema = srClient.getSchemaBySubjectAndId(sourceSubject, id);
-      srClient.register(changelogSubject, schema);
-    } catch (IOException | RestClientException e) {
-      LOG.warn("Failed during deserialization callback for topic " + source, e);
+      if (!failedAttempts.contains(event)) {
+        LOG.info("Trying to fetch & register schema id {} under subject {}", id, changelogSubject);
+        final ParsedSchema schema = srClient.getSchemaBySubjectAndId(sourceSubject, id);
+        srClient.register(changelogSubject, schema);
+      }
+    } catch (Exception e) {
+      LOG.warn("Failed during deserialization callback for topic {}. "
+          + "Will not try again to register id {} under subject {}.",
```

Comment:
To confirm my understanding: if we fail to proactively register a schema and hit the bug, we'd see the same query failure as we do today (inability to deserialize state store record). In other words, we'll have fixed the problem for cases where the source schema is compatible with the ksqlDB changelog schema (assuming no transient failures while registering the new schema), but there is still a gap in the case where the source schema is incompatible?

Reply:
that is not correct - the deserialization will actually succeed (see behavior in https://github.com/confluentinc/ksql/blob/master/ksqldb-serde/src/main/java/io/confluent/ksql/serde/StaticTopicSerde.java) it will just do the following for every message:

1. deserialize under the changelog topic subject
2. fail because schema id doesn't exist
3. deserialize under the source topic subject (succeed)

Registering the schema is an optimization so that step 1 succeeds in the future (prevent us from checking if the schema exists and deserializing the schemaID twice for every message). Perhaps this was a premature optimization, so mea culpa...

Reply:
Got'cha -- love to hear it! Thanks for clarifying.